alloc-fmt/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+//! The macros in this crate are named `alloc_xxx`, where `xxx` is the name of the equivalent
+//! standard library macro (e.g., `alloc_println`, `alloc_debug_assert`, etc).
+//!
+//! The behavior of the formatting macros is identical to the behavior of their standard library
+//! counterparts. The behavior of the assertion macros is somewhat different. When an assertion
+//! fails, a message is first unconditionally printed to stderr (in case further processing causes
+//! a crash). A stack trace is then printed, and the process aborts. If, during the printing of the
+//! stack trace, recursion is detected (computing stack traces can cause allocation, which can
+//! trigger further assertions, etc), the process immediately aborts.
+//!
+//! Unlike the standard library assertion macros, no panic is generated, and once an assertion
+//! failure triggers, it cannot be caught or aborted.
+
+#![no_std]
+#![feature(core_intrinsics)]
+
+extern crate backtrace;
+extern crate libc;
+extern crate spin;
+
+use core::fmt::*;
+use core::sync::atomic::{AtomicBool, ATOMIC_BOOL_INIT};
+
+// Import items that macros need to reference. They will reference them as $crate::foo instead of
+// core::foo since core isn't guaranteed to be imported in the user's scope.
+#[doc(hidden)]
+pub use core::mem::drop;
+#[doc(hidden)]
+pub use core::fmt::Write;
+#[doc(hidden)]
+pub use core::sync::atomic::Ordering::SeqCst;
+
+#[doc(hidden)]
+pub static STDERR_MTX: spin::Mutex<()> = spin::Mutex::new(());
+#[doc(hidden)]
+pub static STDOUT_MTX: spin::Mutex<()> = spin::Mutex::new(());
+
+#[doc(hidden)]
+pub static STDOUT: libc::c_int = 1;
+#[doc(hidden)]
+pub static STDERR: libc::c_int = 2;
+
+#[doc(hidden)]
+pub struct FDWriter(pub libc::c_int);
+
+impl Write for FDWriter {
+    #[inline]
+    fn write_str(&mut self, s: &str) -> Result {
+        let mut buf = s.as_bytes();
+        while !buf.is_empty() {
+            unsafe {
+                let written = libc::write(self.0, buf.as_ptr() as *const _, buf.len());
+                if written < 1 {
+                    core::intrinsics::abort();
+                }
+                buf = &buf[written as usize..];
+            }
+        }
+        Ok(())
+    }
+}
+
+// We can't simply 'pub use core::intrinsics::abort' because its use requires
+// feature(core_intrinsics), which the user would have to enable.
+#[doc(hidden)]
+pub unsafe fn abort() -> ! {
+    core::intrinsics::abort();
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! print_internal {
+    ($file:expr, $mtx:expr, $($arg:tt)*) => {
+        // in case we're called from inside an unsafe block
+        #[allow(unused_unsafe)]
+        unsafe {
+            let guard = $mtx.lock();
+
+            let mut fd = $crate::FDWriter($file);
+            use $crate::Write;
+            let _ = write!(&mut fd, $($arg)*).map_err(|_| {
+                $crate::abort();
+            });
+
+            // explicitly drop the guard so we're guaranteed it lives at least this long;
```

Comment:
Really? That seems wrong... I thought Mutex<()> was a not-uncommon pattern.

Reply:
I'm not positive, but better safe than sorry.